typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1002 @@
+
+
+Let's define interface `Eq`
+
+```typescript
+
+interface Eq<T> {
+  equal: (f: T, s: T) => boolean;
+}
+```
+
+And now two implementations of it: `IntEq` for numbers and `IntArrEq` for arrays of numbers.
+The first one will simply compare two numbers.
+And the second one will return equality if the first array contain all elements from the second array.
+
+```typescript
+class IntEq implements Eq<number> {
+  equal(f: number, s: number) {
+    return f === s;
+  }
+}
+
+class IntArrEq implements Eq<number[]> {
+  equal(f: number[], s: number[]) {
+    return f.filter(item => s.indexOf(item) < 0).length === 0;
+  }
+}
+```
+
+Now let's implement `lookup` function which for key value pairs array and specific key will return the value associated with this key or `undefined`.
+We can write it polymorph passing comparator class which implements `Eq` for the key type.
+```typescript
+const lookup = <T, K extends Eq<T>, V>(cmp: K, k: T, mp: [T, V][]): V | undefined => {
+  let result: V | undefined;
+  mp.forEach(([kk, v]) => {
+    if (cmp.equal(k, kk)) {
+      result = v;
+    }
+  });
+  return result;
+};
+
+lookup(new IntEq(), 1, [[1, "2"], [2, "1"]]); // "2"
+lookup(new IntArrEq(), [1, 2, 3], [[[1, 3], "1"], [[3, 2], "2"], [[2, 1, 3], "3"]]); // "3"
+```
+
+
+#### Row polymorphism
+
+One more polymorphism is row.
+It allows you to make sure that the record contains at least given set of fields.
+Unfortunately TypeScript type compatibility is based on structural subtyping, so what we have available it is not _really_ a row polymorphism.
+Despite this we can use intersection types to emulate this.
+But you should be careful with using it because during assignments additional information can be lost.
+Developer must be cautious to propagate the full type-level information when they need it.
+
+```typescript
+type fnT = <T>(v: T & { x: number }) => T & { x: number }
+// You should fully describe all types to not lose information.
+
+type A = {};
+type B = A & {x: number};
+
+let a: A = {};
+const b: B = {x: 1};
+a = b;
+a.x; // Property 'x' does not exist on type 'A'.
+```
+
+### Mapped types, Conditional types and Type families
+
+One useful type functionality of TypeScript are Mapped and Conditional types.
+They give us really flexible ability to modify existing types or create now one.
+
+#### Mapped types
+
+Mapped types allows you to create new types based on old types transforming them in some way.
+The syntax looks like `[set of field names]: type;`.
+For example, you can add optional modifier or readonly modifier to each field with `type Partial<T> = { [P in keyof T]?: T[P] }` and `type Readonly<T> = { readonly [P in keyof T]: T[P] }`.
+It is also possible to remove such modifiers using `-?` and `-readonly`.
+
+The `keyof` operator in the code below produces a union of properties name.
+Square brackets indicates computed names.
+And `in` operator goes over all field keys.
+
+`Partial` type adds `?` modifier for each property.
```

Comment:
It seems that you now have two examples of the same thing. 

I would just go like: 

* Syntax looks like this. 
* Example 
* Explanation of the syntax
* You can also remove properties
* Example